src/Endpoint/Partition.php

```diff
@@ -0,0 +1,171 @@
+<?php
+namespace Aws\Endpoint;
+
+use InvalidArgumentException as Iae;
+
+/**
+ * @internal
+ */
+final class Partition implements \ArrayAccess
+{
+    private $definition;
+
+    public function __construct(array $definition)
+    {
+        foreach (['partition', 'regions', 'services', 'dnsSuffix'] as $key) {
+            if (!isset($definition[$key])) {
+                throw new Iae("Partition missing required $key field");
+            }
+        }
+
+        $this->definition = $definition;
+    }
+
+    public function toArray()
+    {
+        return $this->definition;
+    }
+
+    public function offsetGet($offset)
+    {
+        return isset($this->definition[$offset])
+            ? $this->definition[$offset] : null;
+    }
+
+    public function offsetSet($offset, $value)
+    {
+        $this->definition[$offset] = $value;
+    }
+
+    public function offsetExists($offset)
+    {
+        return isset($this->definition[$offset]);
+    }
+
+    public function offsetUnset($offset)
+    {
+        unset($this->definition[$offset]);
+    }
+
+    /**
+     * Determine if this partition contains the provided region.
+     *
+     * @param string $region
+     *
+     * @return bool
+     */
+    public function matchesRegion($region)
+    {
+        if (isset($this->definition['regions'][$region])) {
+            return true;
+        }
+
+        if (isset($this->definition['regionRegex'])) {
+            return (bool) preg_match(
+                "@{$this->definition['regionRegex']}@",
+                $region
+            );
+        }
+
+        return false;
+    }
+
+    public function getRegionsForService($service)
+    {
+        if (isset($this->definition['services'][$service]['isRegionalized'])
+            && false === $this->definition['services'][$service]['isRegionalized']
+        ) {
```

Comment:
This first if statement seems to be the same with function resolveRegion($service, $region) despite that array is returned. While I can see that these 2 functions have different purpose, can you resolve the duplicates? Since $region parameter is only returned when the if check fails?